File: app/integrations/webhook_signing.py
from __future__ import annotations

import hmac
import hashlib
from typing import Optional


def _safe_str_eq(a: str, b: str) -> bool:
    return hmac.compare_digest(a.encode("utf-8"), b.encode("utf-8"))


def sign_hmac_sha256(secret: str, body: bytes) -> str:
    mac = hmac.new(secret.encode("utf-8"), body, hashlib.sha256)
    return mac.hexdigest()


def verify_hmac_sha256(
    *,
    secret: str,
    body: bytes,
    signature: Optional[str],
) -> bool:
    if not signature:
        return False
    expected = sign_hmac_sha256(secret, body)
    return _safe_str_eq(expected, signature)


# DEAD (currently unused): alternative signature format, not referenced anywhere
def verify_hmac_sha256_prefixed(
    *,
    secret: str,
    body: bytes,
    signature: Optional[str],
    prefix: str = "sha256=",
) -> bool:
    if not signature or not signature.startswith(prefix):
        return False
    return verify_hmac_sha256(secret=secret, body=body, signature=signature[len(prefix) :])

File: web/src/integrations/webhookSigning.ts
// web/src/integrations/webhookSigning.ts
import crypto from "crypto";

function safeStrEq(a: string, b: string): boolean {
  const bufA = Buffer.from(a, "utf-8");
  const bufB = Buffer.from(b, "utf-8");
  if (bufA.length !== bufB.length) 
    return false;
  return crypto.timingSafeEqual(bufA, bufB);
}

export function signHmacSha256(secret: string, body: Buffer): string {
  return crypto.createHmac("sha256", secret).update(body).digest("hex");
}

export function verifyHmacSha256(secret: string, body: Buffer, signature: string | undefined): boolean {
  if (!signature) 
    return false;
  const expected = signHmacSha256(secret, body);
  return safeStrEq(expected, signature);
}

// UNUSED (demo): alternative signature format, not referenced anywhere
export function verifyHmacSha256Prefixed(
  secret: string,
  body: Buffer,
  signature: string | undefined,
  prefix: string = "sha256="
): boolean {
  if (!signature || !signature.startsWith(prefix)) 
    return false;
  return verifyHmacSha256(secret, body, signature.slice(prefix.length));
}
